### The time it takes for a computer program to do a certain task before and after an improvement in the program code is given below. Find the 95% confidence interval for the difference between the average running time of the program before and after the improvement and test whether the improvement in the program code worked.

In [6]:
library(readxl)
library(tidyverse)

In [2]:
df <- read_excel("/data/notebook_files/Software_Data.xlsx")

In [4]:
head(df)

A tibble: 6 × 2 
 
 Before After 
 <chr> <chr> 
 
 
 26.2 25.0 
 29.3 28.2 
 30.2 28.2 
 32.1 31.8 
 24.0 24.0 
 32.0 31.0

In [7]:
glimpse(df)

Rows: 8
Columns: 2
$ Before <chr> "26.2", "29.3", "30.2", "32.1", "24.0", "32.0", "27.6", "26.5"
$ After  <chr> "25.0", "28.2", "28.2", "31.8", "24.0", "31.0", "26.5", "25.9"


In [13]:
df$Before <- as.numeric(df$Before)
df$After <- as.numeric(df$After)
df$D <- df$Before - df$After
glimpse(df)

Rows: 8
Columns: 3
$ Before <dbl> 26.2, 29.3, 30.2, 32.1, 24.0, 32.0, 27.6, 26.5
$ After  <dbl> 25.0, 28.2, 28.2, 31.8, 24.0, 31.0, 26.5, 25.9
$ D      <dbl> 1.2, 1.1, 2.0, 0.3, 0.0, 1.0, 1.1, 0.6


#### Normality Analysis

                    𝐻0: There is no difference between the distribution of the data and the normal distribution.

                    𝐻𝑠: There is a difference between the distribution of the data and the normal distribution.

In [15]:
shapiro.test(df$D)


	Shapiro-Wilk normality test

data:  df$D
W = 0.95205, p-value = 0.7318


(p=0.7318) > 0.05, 𝐻0 cannot be rejected. The distribution of the data on the difference between the time it takes the program to do the job before and after the improvement shows a normal distribution.

#### Dependent t-test
𝑛<30, a t-test is used to test the difference between two dependent population means.

In [16]:
t.test(df$D,mu=0, alternative="two.sided", conf.level=0.95)


	One Sample t-test

data:  df$D
t = 4.1957, df = 7, p-value = 0.004057
alternative hypothesis: true mean is not equal to 0
95 percent confidence interval:
 0.3982331 1.4267669
sample estimates:
mean of x 
   0.9125 


- (p=0.004) < 0.05 -> p value is less than the significance value (𝛼=0.05)
  
- |t|=4.1957 test statistic 𝑡(𝛼/2,𝑛-1) = 𝑡(0.025,7) = 2.365 is greater than the table value.

- (0.398, 1.426) confidence interval does not include 0.

#### Conclusion

According to the t-test result, the hypothesis 𝐻0 is rejected. It can be said that there is a difference at 5% significance level between the duration of doing the job before and after making improvements to the program. The improvement worked and the time to do the job of the program after the improvement was made was shortened.